File: 11.zig
//! https://adventofcode.com/2023/day/11

const std = @import("std");
const Pt = @Vector(2, isize);
const Map = std.AutoArrayHashMap(Pt, void);

pub fn main() !void {
    var arena = std.heap.ArenaAllocator.init(std.heap.page_allocator);
    defer arena.deinit();
    const alloc = arena.allocator();
    const args = try std.process.argsAlloc(alloc);
    const file = try std.fs.cwd().openFile(args[1], .{});
    defer file.close();
    const input = try file.readToEndAlloc(alloc, std.math.maxInt(u32));

    var lines = std.mem.tokenizeScalar(u8, input, '\n');
    var map = Map.init(alloc);
    var extents: Pt = .{ 0, 0 };
    // parse
    var y: isize = 0;
    while (lines.next()) |line| : (y += 1) {
        var x: isize = 0;
        for (line) |c| {
            if (c == '#') try map.put(.{ x, y }, {});
            x += 1;
        }
        extents[0] = x;
    }
    extents[1] = y;

    var map2 = try map.clone();
    var extents2 = extents;
    const part1 = try solve(&map, &extents, 1);
    const part2 = try solve(&map2, &extents2, 999_999);
    std.debug.print("part1 {} part2 {}\n", .{ part1, part2 });
    std.debug.assert(part1 == 10289334);
    std.debug.assert(part2 == 649862989626);
}

fn solve(map: *Map, extents: *Pt, comptime expansion: comptime_int) !usize {
    { // vertical expansion
        var y: isize = 0;
        while (y < extents[1]) : (y += 1) {
            const any = blk: {
                var x: isize = 0;
                while (x < extents[0]) : (x += 1) {
                    if (map.contains(.{ x, y })) break :blk true;
                }
                break :blk false;
            };
            if (!any) {
                for (map.keys()) |*pt| {
                    if (pt[1] > y) pt[1] += expansion;
                }
                y += expansion;
                extents[1] += expansion;
                try map.reIndex();
            }
        }
    }
    { // horizontal expansion
        var x: isize = 0;
        while (x < extents[0]) : (x += 1) {
            const any = for (map.keys()) |k| {
                if (k[0] == x) break true;
            } else false;
            if (!any) {
                for (0..map.keys().len) |i| {
                    const pt = &map.keys()[i];
                    if (pt[0] > x) pt[0] += expansion;
                }
                x += expansion;
                extents[0] += expansion;
                try map.reIndex();
            }
        }
    }
    var sum: usize = 0;
    for (1..map.count()) |i| {
        const first = map.keys()[i - 1];
        for (map.keys()[i..]) |next|
            sum += @reduce(.Add, @abs(first - next));
    }
    return sum;
}

fn drawMap(map: Map, extents: Pt) void {
    // std.debug.print("extents={any}\n", .{extents});
    var y: isize = 0;
    while (y < extents[1]) : (y += 1) {
        var x: isize = 0;
        while (x < extents[0]) : (x += 1) {
            if (map.contains(.{ x, y })) {
                std.debug.print("#", .{});
            } else std.debug.print(".", .{});
        }
        std.debug.print("\n", .{});
    }
}

File: 11.rs
use std::env;

fn main() -> Result<(), ()> {
    let args = env::args();
    let filepath = args.skip(1).next().ok_or(())?;
    let s = std::fs::read(filepath).or(Err(()))?;
    let input = String::from_utf8_lossy(&s);
    let mut map = Vec::<[usize; 2]>::new();
    let extents = [
        input.bytes().position(|c| c == b'\n').ok_or(())?,
        input.bytes().fold(0, |acc, c| acc + (c == b'\n') as usize),
    ];

    for (y, line) in input.split("\n").filter(|s| !s.is_empty()).enumerate() {
        for (x, c) in line.chars().enumerate() {
            if c == '#' {
                map.push([x, y]);
            }
        }
    }

    let mut map2 = map.clone();
    let part1 = solve(&mut map, extents, 1);
    let part2 = solve(&mut map2, extents, 999_999);
    println!("part1 {part1} part2 {part2}");
    assert!(part1 == 10289334);
    assert!(part2 == 649862989626);
    Ok(())
}

fn adjust(map: &mut Vec<[usize; 2]>, extents: &mut [usize; 2], expansion: usize, idx: usize) {
    let mut coord: usize = 0;
    while coord < extents[idx] {
        if map.iter().find(|pt| pt[idx] == coord).is_none() {
            for pt in map.iter_mut() {
                if pt[idx] > coord {
                    pt[idx] += expansion;
                }
            }
            coord += expansion;
            extents[idx] += expansion;
        }
        coord += 1;
    }
}

fn solve(map: &mut Vec<[usize; 2]>, extents: [usize; 2], expansion: usize) -> usize {
    let mut extents = extents;
    adjust(map, &mut extents, expansion, 0);
    adjust(map, &mut extents, expansion, 1);

    let mut sum: usize = 0;
    for i in 1..map.len() {
        let first = map[i - 1];
        for next in &map[i..] {
            sum += ((first[0] as isize - next[0] as isize).abs()
                + (first[1] as isize - next[1] as isize).abs()) as usize;
        }
    }
    sum
}
